dali/operators/reader/loader/video/frames_decoder.h

```diff
@@ -0,0 +1,190 @@
+  int stream_id_ = -1;
+
+  ~AvState() {
+    sws_freeContext(sws_ctx_);
+    av_packet_free(&packet_);
+    av_frame_free(&frame_);
+    avcodec_free_context(&codec_ctx_);
+    avformat_close_input(&ctx_);
+    avformat_free_context(ctx_);
+
+    ctx_ = nullptr;
+    codec_ = nullptr;
+    codec_params_ = nullptr;
+    codec_ctx_ = nullptr;
+    frame_ = nullptr;
+    packet_ = nullptr;
+    sws_ctx_ = nullptr;
+  }
+};
+
+/**
+ * @brief Object representing a video file. Allows access to frames and seeking.
+ * 
+ */
+class DLL_PUBLIC FramesDecoder {
+ public:
+
+  /**
+   * @brief Construct a new FramesDecoder object.
+   * 
+   * @param filename Path to a video file.
+   */
+  FramesDecoder(const std::string &filename);
+
+  /**
+   * @brief Number of frames in the video
+   * 
+   */
+  int64_t NumFrames() const {
+    return index_.size();
+  }
+
+  /**
+   * @brief Width of a video frame in pixels
+   * 
+   */
+  int Width() const {
+    return av_state_->codec_params_->width;
+  }
+
+  /**
+   * @brief Height of a video frame in pixels
+   * 
+   */
+  int Height() const {
+    return av_state_->codec_params_->height;
+  }
+
+  /**
+   * @brief Number of channels in a video
+   * 
+   */
+  int Channels() const {
+      return channels_;
+  }
+
+  /**
+   * @brief Total number of values in a frame (width * height * channels)
+   * 
+   */
+  int FrameSize() const {
+    return Channels() * Width() * Height();
+  }
+
+  /**
+   * @brief Reads next frame of the video and copies it to the provided buffer, if copy_to_output is True.
+   * 
+   * @param data Output buffer to copy data to.
+   * @param copy_to_output Whether copy the data to the output. 
+   * @return Boolean indicating whether the frame was read or not. False means no more frames in the decoder.
+   */
+  bool ReadNextFrame(uint8_t *data, bool copy_to_output = true);
+
+  /**
+   * @brief Seeks to the frame given by id. Next call to ReadNextFrame will return this frame
+   * 
+   * @param frame_id Id of the frame to seek to
+   */
+  void SeekFrame(int frame_id);
+
+  /**
+   * @brief Seeks to the first frame
+   * 
+   */
+  void Reset();
+
+  DISABLE_COPY_MOVE_ASSIGN(FramesDecoder);
+
+ private:
+   /**
+   * @brief Gets the packet from the decoder and reads a frame from it to provided buffer. Returns 
+   * boolean indicating, if the frame was succesfully read.
+   * After this method returns false, there might be more frames to read. Call `ReadFlushFrame` until
+   * it returns false, to get all of the frames from the video file.
+   * 
+   * @param data Output buffer to copy data to. If `copy_to_output` is false, this value is ignored.
+   * @param copy_to_output Whether copy the frame to provided output.
+   * 
+   * @returns True, if the read was succesful, or false, when all regular farmes were consumed.
+   * 
+   */
+  bool ReadRegularFrame(uint8_t *data, bool copy_to_output = true);
+
+  /**
+   * @brief Reads frames from the last packet. This packet can hold
```

Comment:
Is it about the packet holding multiple frames, or the decoder that receiving the last packet has enough context to generate multiple frames?